src/ndb/ndb_internal.c

```diff
@@ -0,0 +1,426 @@
+      break;
+    default:
+      assert(true);
+      break;
+  }
+}
+
+void ndb_load_data(ndb_file_t *f,
+                   const char* ftype,
+                   void *data,
+                   ndb_element_metadata_t* metadata,
+                   const size_t el_size,
+                   const size_t el_number)
+{
+  enum ndb_op_code r = NDB_ERR_FILE_IO;
+  size_t ds = el_size * el_number;
+  size_t mds = sizeof(ndb_element_metadata_nv_t) * el_number;
+  ndb_element_metadata_nv_t md_nv[el_number];
+
+  if (do_writes_to_flash) {
+    if (ndb_open_file(f) == NDB_ERR_NONE) {
+      if (ndb_read(f, 0, data, ds) == NDB_ERR_NONE) {
+        if (ndb_read(f, ds, md_nv, mds) == NDB_ERR_NONE) {
+          u32 i;
+          for (i = 0; i < el_number; i++) {
+            metadata[i].nv_data = md_nv[i];
+          }
+          r = NDB_ERR_NONE;
+        }
+      }
+    }
+  }
+
+  memset(data, 0, ds);
+  memset(metadata, 0, sizeof(ndb_element_metadata_t) * el_number);
+
+  if (do_writes_to_flash && (f->fh >= 0)) {
+    memset(md_nv, 0, mds);
+    if (ndb_write_file_data(f, 0, data, ds) == NDB_ERR_NONE) {
+      if (ndb_write_file_data(f, ds, md_nv, mds) == NDB_ERR_NONE) {
+        if (ndb_write_file_data(f, ds + mds,
+                                &ndb_file_end_mark, 1) == NDB_ERR_NONE) {
+          r = NDB_ERR_INIT_DONE;
+        }
+      }
+    }
+  }
+
+  ndb_log_file_open(r, ftype, f);
+
+  for (u32 i = 0; i < el_number; i++) {
+    metadata[i].data = data + (i * el_size);
+    metadata[i].index = i;
+    metadata[i].file = f;
+    metadata[i].next = NULL;
+    metadata[i].update_c = 1;
+  }
+}
+
+ndb_timestamp_t ndb_get_timestamp()
+{
+  return chVTGetSystemTime();
+}
+
+static void ndb_service_thread(void* p)
+{
+  (void) (p);
+  chRegSetThreadName("ndb");
+  bool data_write_ok = false;
+  bool md_write_ok = false;
+
+  while (true) {
+    if (NDB_ERR_TIMEOUT == do_nv_writes(&data_write_ok, &md_write_ok)) {
+      do_sbp_updates();
+    } else {
+      if (!do_writes_to_flash)
+        continue;
+      if (!data_write_ok)
+        log_error("Error writing data to the NDB file");
+      if (!md_write_ok)
+        log_error("Error writing metadata to the NDB file");
+    }
+  }
+}
+
+enum ndb_op_code do_nv_writes(bool *data_write_ok, bool *md_write_ok)
+{
+  ndb_element_metadata_t* md = NULL;
+  enum ndb_op_code ret;
+  int changed;
+
+  if (NDB_ERR_TIMEOUT == ndb_wq_get(&md)) {
+    return NDB_ERR_TIMEOUT;
+  } else {
+    if (!do_writes_to_flash) {
+      *data_write_ok = *md_write_ok = true;
+      return NDB_ERR_NONE;
+    }
+    ndb_element_metadata_t md_copy;
+    u16 element_size = md->file->data_size;
+    u16 n_elements = md->file->n_elements;
+    u8 buf[element_size];
+    ndb_lock();
+
+    assert((md->nv_data.state & (NDB_IE_DIRTY + NDB_MD_DIRTY)) != 0);
+
+    if(md->nv_data.state & NDB_IE_DIRTY)
+      memcpy(&buf, md->data, element_size);
+
+    memcpy(&md_copy, md, sizeof(ndb_element_metadata_t));
```

Comment:
I had that initially, but then thought that as metadata are to be copied and saved always, that check would be redundant. Even NDB_MD_DIRTY flag can be dropped. I just want to keep it so that there is clear indication what caused the write request.